autoload/dirvish.vim

```diff
@@ -466,6 +482,44 @@ function! dirvish#open(...) range abort
   call s:open_dir(d, reloading)
 endfunction
 
+function! dirvish#add_column(column) abort
+  for l:col in s:columns
+    if a:column.name ==? l:col.name
+      return
+    endif
+  endfor
+
+  call add(s:columns, a:column)
+  return sort(s:columns, 's:sort_columns')
+endfunction
+
+function! dirvish#get_columns() abort
+  return copy(s:columns)
+endfunction
+
+function! s:get_column_for_path(path) abort
+  for l:column in s:columns
+    let l:should_mark = call(l:column.handler, [a:path])
+    if l:should_mark
+      return l:column.mark
+    endif
+  endfor
+  return ''
+endfunction
+
+function! s:strip_columns(path) abort
+  let l:marks = join(map(copy(s:columns), 'escape(v:val.mark, "*.^$~\\")'), '\|')
+  return substitute(a:path, '^\('.l:marks.'\)', '', '')
```

Comment:
IIUC you are inserting the mark in the buffer. 

Can't conceal be used instead? Probably difficult to get the syntax definition correct. But one of Dirvish's main design goals is to avoid extra junk in the buffer contents. (Whitespace is probably a reasonable exception.)